trail_detection_node/trail_detection_node/v2_trailDetectionNode.py

```diff
@@ -12,241 +12,635 @@
+        """
+
+        # Use camera properties to convert pixels to uv
+        uv_route = np.empty_like(route_indices).astype(float) #N, 2 
+        uv_route[:, 0] = (route_indices[:, 0] - self.c_x) / self.f_x
+        uv_route[:, 1] = (route_indices[:, 1] - self.c_y) / self.f_y
+        return uv_route
+    
+    def compute_route_uv_and_img(self, camera_msg: Image) -> tuple[np.ndarray, np.ndarray]:
+        """
+        Computes the planned route in uv coordinates and produces the image for visualization.
+
+        Args:
+            camera_msg (sensor_msgs.msg.Image): Camera message.
+
+        Returns:
+            tuple: Tuple containing route in uv coordinates(numpy.ndarray) and image with trail and centreline (numpy.ndarray).
+        """
+
+        def get_rgb_undistorted_img(camera_msg: Image) -> np.ndarray:
+            """
+            Processes the camera message to retrieve the undistorted RGB image.
+
+            Args:
+                camera_msg (sensor_msgs.msg.Image): Camera message.
+
+            Returns:
+                numpy.ndarray: Undistorted RGB image.
+            """
+            # process camera msg and set to RGB
+            cv_image = self.bridge.imgmsg_to_cv2(camera_msg, desired_encoding='passthrough')
+            cv_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2RGB)
+
+            # Undistort image
+            if not self.image_height:
+                self.image_height, self.image_width = cv_image.shape[:2]
+            new_K, _ = cv2.getOptimalNewCameraMatrix(CAM_INTRINSIC_K, RADIAL_DISTORTION, (self.image_width, self.image_height), 1, (self.image_width, self.image_height))  
+            cv_image = cv2.undistort(cv_image, CAM_INTRINSIC_K, RADIAL_DISTORTION, None, new_K)
+            return cv_image
+
+        def pre_process_img(image: np.ndarray, pre_proc_blur_k_size: int, brightness: int) -> np.ndarray:
+            """
+            Pre-processes the image for segmentation model.
+            Applies equalization, Gaussian Blur, and brightness increase.
+
+            Args:
+                image (numpy.ndarray): RGB undistorted input image.
+                pre_proc_blur_k_size (int): Kernel size for Gaussian blur.
+                brightness (int): Brightness adjustment value.
+
+            Returns:
+                numpy.ndarray: Pre-processed image.
+            """
+
+            # Split the image into its color channels
+            r, g, b = cv2.split(image)
+
+            # Equalize the histograms for each channel
+            r_eq = cv2.equalizeHist(r)
+            g_eq = cv2.equalizeHist(g)
+            b_eq = cv2.equalizeHist(b)
+
+            # Merge the channels
+            image = cv2.merge((r_eq, g_eq, b_eq))
+
+            # Apply Gaussian Blur
+            image = cv2.GaussianBlur(image, (pre_proc_blur_k_size, pre_proc_blur_k_size), 0)
+
+            # # Increase Brightness
+            M = np.ones(image.shape, dtype='uint8') * brightness  # Increase brightness by 50 
```

Comment:
Yeah that's a good point, I had not considered the brightness above 255, I can adjust the code to ensure that doesn't occur. I can try the Gamma transform instead if you think that would be more appropriate. From my understanding, the histogram equalization deals more with the contrast, whereas gamma transformation deals more with brightness. And the goal of this step is to make the features more noticeable in darker lighting.

Reply:
In practice, it might not be a problem, but it's still not ideal. The problem with capping the values is that this will destroy features, and so that's probably worse than having larger than expected inputs. You could have a look at gamma correction, but I think we should just test a few things on some images, but it might not be critical for now.